SynthText dataset integration

felixdittrich92 opened the pull request:
This PR adds the Synthtext dataset 

cf. #587

Problems:

- the unpack part needs a bit longer  (>800k data) so i think we have to implement a check that unpacking is finished before continue
- same for the hash check this will end in a MemoryError

wdyt @fg-mindee how to handle? :)
suggestions are welcome :hugs: 

- [x] tests
- [x] bypass the hash check

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+
+    def __init__(
+        self,
+        train: bool = True,
```
There is no train/val split?
If so, we should make it ourselves (say the first 90% goes into train, 10% for val)

felixdittrich92 replied:
```
if train:
            self.data = self.data[:int(len(self.data) * 0.9)]
        else:
            self.data = self.data[int(len(self.data) * 0.9):]
```
@fg-mindee 
ok ?

fg-mindee replied:
well let's avoid unnecessary iterations:
- if train --> at 90% we break the iteration
- if val --> we start iterating only at 90%

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+
+    Args:
+        sample_transforms: composable transformations that will be applied to each image
+        rotated_bbox: whether polygons should be considered as rotated bounding box (instead of straight ones)
```
missing arg description for train

fg-mindee replied:
still missing

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+            if not os.path.exists(os.path.join(tmp_root, img_path[0])):
+                raise FileNotFoundError(f"unable to locate {os.path.join(tmp_root, img_path[0])}")
+
+            labels = _text_to_words(txt)
```
let's avoid other functions when it's not needed:
```python
labels = txt.tolist()
```

felixdittrich92 replied:
@fg-mindee does not work correctly but now:
`labels = '\n'.join(txt).split()`

felixdittrich92 replied:
ok ?

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+        self.root = tmp_root
+
+
+def _text_to_words(txt: np.ndarray) -> List[str]:
+    """Convert np.str-Array to list of str."""
+    line = '\n'.join(txt)
+    return line.split()
```
can be removed

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+                box_targets = [_compute_rotated_box(pts) for pts in word_boxes]
+            else:
+                # xmin, ymin, xmax, ymax
+                box_targets = [_compute_straight_box(pts) for pts in word_boxes]  # type: ignore[misc]
```
no need:
```python
box_targets = np.concatenate((word_boxes.min(axis=1), word_boxes.max(axis=1)), axis=1)
```

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+
+            if rotated_bbox:
+                # x_center, y_center, w, h, alpha = 0
+                box_targets = [_compute_rotated_box(pts) for pts in word_boxes]
```
Better to vectorize this:
```python
mins = word_boxes.min(axis=1)
maxs = word_boxes.max(axis=1)
box_targets = np.concatenate(((mins + maxs) / 2, maxs - mins, np.zeros(word_boxes.shape[0])), axis=1)
```

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,103 @@
+    return line.split()
+
+
+def _compute_straight_box(pts: np.ndarray) -> Tuple[float, float, float, float]:
+    # pts: Nx2
+    xmin = np.min(pts[:, 0])
+    xmax = np.max(pts[:, 0])
+    ymin = np.min(pts[:, 1])
+    ymax = np.max(pts[:, 1])
+    return xmin, ymin, xmax, ymax
+
+
+def _compute_rotated_box(pts: np.ndarray) -> Tuple[float, float, float, float, int]:
+    # pts: Nx2
+    x = np.min(pts[:, 0])
+    y = np.min(pts[:, 1])
+    width = np.max(pts[:, 0]) - x
+    height = np.max(pts[:, 1]) - y
+    # x_center, y_center, w, h, alpha = 0
+    return x + width / 2, y + height / 2, width, height, 0
```
Can be removed

fg-mindee commented on doctr/utils/data.py:
```diff
@@ -37,10 +37,14 @@ def _urlretrieve(url: str, filename: Union[Path, str], chunk_size: int = 1024) -
-    with open(file_path, 'rb') as f:
-        sha_hash = hashlib.sha256(f.read()).hexdigest()
+    try:
+        with open(file_path, 'rb') as f:
+            sha_hash = hashlib.sha256(f.read()).hexdigest()
 
-    return sha_hash[:len(hash_prefix)] == hash_prefix
+        return sha_hash[:len(hash_prefix)] == hash_prefix
+    # FIX: for big datasets
+    except MemoryError:
+        return True
```
We can't afford that, if the dataset causes hash issues, we can consider avoiding the hash check, but this function should not be modified

fg-mindee reviewed (changes requested):
Thanks! I added some comments

felixdittrich92 commented:
@fg-mindee 
Dataset size: ~41 GB

fg-mindee reviewed (commented):
Another round of comments :sweat_smile:

felixdittrich92 commented:
@fg-mindee 
I added to bypass the hash check and removed the tests for this dataset ftm also the train/val split now before iterating.
Already any decision about the dataset test ? :)

felixdittrich92 commented:
@fg-mindee 
after attaching the example can you trigger integration again pls ? :smiley:

fg-mindee commented on doctr/utils/data.py:
```diff
@@ -49,6 +49,7 @@ def download_from_url(
     hash_prefix: Optional[str] = None,
     cache_dir: Optional[str] = None,
     cache_subdir: Optional[str] = None,
+    check_hash: bool = True,
```
As mentioned before, we can't forward to change this file :sweat_smile: 
If at some point, we cannot perform the check, then we just don't do it upstream but this function should not be changed. If the name of the file doesn't include any hash and it's not passed, the check won't be done, so that should be enough :ok_hand:

fg-mindee commented on doctr/datasets/datasets/base.py:
```diff
@@ -89,7 +91,7 @@ def __init__(
         if not os.path.exists(archive_path) and not download:
             raise ValueError("the dataset needs to be downloaded first with download=True")
 
-        archive_path = download_from_url(url, file_name, file_hash, cache_subdir='datasets')
+        archive_path = download_from_url(url, file_name, file_hash, check_hash=check_hash, cache_subdir='datasets')
```
cf my other comment, let's revert the changes on this file :pray:

fg-mindee commented on tests/tensorflow/test_datasets_tf.py:
```diff
@@ -8,26 +8,28 @@
+        ['SynthText', False, [512, 512], 3, True, False],
     ],
 )
-def test_dataset(dataset_name, train, input_size, size, rotate):
+def test_dataset(dataset_name, train, input_size, size, rotate, download):
 
     ds = datasets.__dict__[dataset_name](
-        train=train, download=True, sample_transforms=Resize(input_size), rotated_bbox=rotate,
+        train=train, download=download, sample_transforms=Resize(input_size), rotated_bbox=rotate,
```
cf the comments above

fg-mindee commented on tests/pytorch/test_datasets_pt.py:
```diff
@@ -18,26 +18,28 @@ def test_visiondataset():
+        ['SROIE', True, [512, 512], 626, False, True],
+        ['SROIE', False, [512, 512], 360, False, True],
+        ['CORD', True, [512, 512], 800, True, True],
+        ['CORD', False, [512, 512], 100, False, True],
+        ['DocArtefacts', True, [512, 512], 2700, False, True],
+        ['DocArtefacts', False, [512, 512], 300, True, True],
+        ['IIIT5K', True, [32, 128], 2000, True, True],
+        ['IIIT5K', False, [32, 128], 3000, False, True],
+        ['SVT', True, [512, 512], 100, True, True],
+        ['SVT', False, [512, 512], 249, False, True],
+        ['SynthText', True, [512, 512], 27, False, False],
+        ['SynthText', False, [512, 512], 3, True, False],
     ],
 )
-def test_dataset(dataset_name, train, input_size, size, rotate):
+def test_dataset(dataset_name, train, input_size, size, rotate, download):
 
     ds = datasets.__dict__[dataset_name](
-        train=train, download=True, sample_transforms=Resize(input_size), rotated_bbox=rotate,
+        train=train, download=download, sample_transforms=Resize(input_size), rotated_bbox=rotate,
```
better to do an exception to create the `ds` object like
```python
if dataset_name.lower() == "synthtext":
...
```

The idea here is the following:
- if it's synthtext, we set `download=False`
- before calling the constructor, we download the subsample, put it in the cache
- when we call the constructor, we need to monkeypatch the expected hash

felixdittrich92 replied:
@fg-mindee 
Where to download the subsample (URL) ?? :sweat_smile: 
Can you take a look at the monkeypatch i have never used / needed this before :smiley:

fg-mindee reviewed (commented):
Thanks for the edits! I suggest waiting a bit, because it's gonna be troublesome to get the CI to work nicely with this

felixdittrich92 commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,84 @@
+    """
+
+    URL = 'https://thor.robots.ox.ac.uk/~vgg/data/scenetext/SynthText.zip'
+    #SHA256 = '28ab030485ec8df3ed612c568dd71fb2793b9afbfa3a9d9c6e792aef33265bf1'
```
@fg-mindee 
is it ok to hold this as comment ?

fg-mindee replied:
It would be preferable to uncomment this. But you mentioned the file was too big to be hash?

felixdittrich92 replied:
yes, i would just like to leave it there in case someone comes up with the idea of ​​wanting to check it out

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,85 @@
+        np_dtype = np.float32
+
+        for img_path, word_boxes, txt in tqdm(iterable=zip(paths, boxes, labels),
+                                              desc='Load SynthText', total=len(paths)):
```
"Loading SynthText..."

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,85 @@
+                mins = word_boxes.min(axis=1)
+                maxs = word_boxes.max(axis=1)
+                box_targets = np.concatenate(
+                    ((mins + maxs) / 2, maxs - mins, np.expand_dims(np.zeros(word_boxes.shape[0]), axis=1)), axis=1)
```
You can replace `np.expand_dims(np.zeros(word_boxes.shape[0]), axis=1)` with `np.zeros((word_boxes.shape[0], 1))`

fg-mindee commented on tests/pytorch/test_datasets_pt.py:
```diff
@@ -32,12 +33,25 @@ def test_visiondataset():
         ['IIIT5K', False, [32, 128], 3000, False],
         ['SVT', True, [512, 512], 100, True],
         ['SVT', False, [512, 512], 249, False],
+        ['SynthText', True, [512, 512], 27, True],  # org: 772875
+        ['SynthText', False, [512, 512], 3, False],  # org: 85875
```
"org: " isn't very explicit
--> "Actual set has XXX samples"

fg-mindee commented on tests/pytorch/test_datasets_pt.py:
```diff
@@ -32,12 +33,25 @@ def test_visiondataset():
     ],
 )
 def test_dataset(dataset_name, train, input_size, size, rotate):
 
+    if dataset_name.lower() == 'synthtext':
+        download = False
+
+        # Download the subsample of SynthText dataset and save in cache
+        URL = 'https://github.com/mindee/doctr/releases/download/v0.4.1/synthtext_samples-89fd1445.zip'
+        FILE_HASH = '89fd1445457b9ad8391e17620c6ae1b45134be2bf5449f36e7e4275176cc16ac'
+        FILE_NAME = 'SynthText.zip'
+        download_from_url(URL, FILE_NAME, FILE_HASH, cache_subdir='datasets')
+    else:
+        download = True
+
     ds = datasets.__dict__[dataset_name](
-        train=train, download=True, sample_transforms=Resize(input_size), rotated_bbox=rotate,
+        train=train, download=download, sample_transforms=Resize(input_size), rotated_bbox=rotate,
```
You can revert this

fg-mindee commented on tests/pytorch/test_datasets_pt.py:
```diff
@@ -32,12 +33,25 @@ def test_visiondataset():
 )
 def test_dataset(dataset_name, train, input_size, size, rotate):
 
+    if dataset_name.lower() == 'synthtext':
+        download = False
+
+        # Download the subsample of SynthText dataset and save in cache
+        URL = 'https://github.com/mindee/doctr/releases/download/v0.4.1/synthtext_samples-89fd1445.zip'
+        FILE_HASH = '89fd1445457b9ad8391e17620c6ae1b45134be2bf5449f36e7e4275176cc16ac'
+        FILE_NAME = 'SynthText.zip'
+        download_from_url(URL, FILE_NAME, FILE_HASH, cache_subdir='datasets')
+    else:
+        download = True
+
```
```python
if dataset_name.lower() == "synthtext":
    # Monkeypatch the class to download a subsample
    datasets.__dict__[dataset_name].URL = 'https://github.com/mindee/doctr/releases/download/v0.4.1/synthtext_samples-89fd1445.zip'
    datasets.__dict__[dataset_name].SHA256 = '89fd1445457b9ad8391e17620c6ae1b45134be2bf5449f36e7e4275176cc16ac'
```

fg-mindee replied:
(we really don't need the FILE_NAME)

fg-mindee commented on tests/tensorflow/test_datasets_tf.py:
```diff
@@ -22,12 +23,25 @@
 def test_dataset(dataset_name, train, input_size, size, rotate):
 
+    if dataset_name.lower() == 'synthtext':
+        download = False
+
+        # Download the subsample of SynthText dataset and save in cache
+        URL = 'https://github.com/mindee/doctr/releases/download/v0.4.1/synthtext_samples-89fd1445.zip'
+        FILE_HASH = '89fd1445457b9ad8391e17620c6ae1b45134be2bf5449f36e7e4275176cc16ac'
+        FILE_NAME = 'SynthText.zip'
+        download_from_url(URL, FILE_NAME, FILE_HASH, cache_subdir='datasets')
+    else:
+        download = True
+
     ds = datasets.__dict__[dataset_name](
-        train=train, download=True, sample_transforms=Resize(input_size), rotated_bbox=rotate,
+        train=train, download=download, sample_transforms=Resize(input_size), rotated_bbox=rotate,
```
same here

fg-mindee commented on tests/tensorflow/test_datasets_tf.py:
```diff
@@ -22,12 +23,25 @@
         ['IIIT5K', False, [32, 128], 3000, False],
         ['SVT', True, [512, 512], 100, True],
         ['SVT', False, [512, 512], 249, False],
+        ['SynthText', True, [512, 512], 27, True],  # org: 772875
+        ['SynthText', False, [512, 512], 3, False],  # org: 85875
```
same here

fg-mindee commented on tests/tensorflow/test_datasets_tf.py:
```diff
@@ -5,6 +5,7 @@
 from doctr import datasets
 from doctr.datasets import DataLoader
 from doctr.transforms import Resize
+from doctr.utils.data import download_from_url
```
can be removed

fg-mindee commented on tests/pytorch/test_datasets_pt.py:
```diff
@@ -5,6 +5,7 @@
 
 from doctr import datasets
 from doctr.transforms import Resize
+from doctr.utils.data import download_from_url
```
can be removed

fg-mindee reviewed (changes requested):
Thanks for taking a look at this Felix!
I added some modification suggestions :ok_hand:

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,85 @@
+        tmp_root = os.path.join(self.root, 'SynthText')
+        mat_data = sio.loadmat(os.path.join(tmp_root, 'gt.mat'))
+        split = int(len(mat_data['imnames'][0]) * 0.9)
+        paths = mat_data['imnames'][0][:split] if self.train else mat_data['imnames'][0][split:]
+        boxes = mat_data['wordBB'][0][:split] if self.train else mat_data['wordBB'][0][split:]
+        labels = mat_data['txt'][0][:split] if self.train else mat_data['txt'][0][split:]
```
Let's keep things simple:
```python
paths = mat_data['imnames'][0][:split if self.train else split:]
boxes = mat_data['wordBB'][0][:split if self.train else split:]
labels = mat_data['txt'][0][:split if self.train else split:]
```

felixdittrich92 replied:
yeah thats right :)

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,85 @@
+            if not os.path.exists(os.path.join(tmp_root, img_path[0])):
+                raise FileNotFoundError(f"unable to locate {os.path.join(tmp_root, img_path[0])}")
+
+            labels = '\n'.join(txt).split()
```
why are we joining only to split?
Can't we split + strip directly?

felixdittrich92 replied:
Mhh no `txt ` is a really ugly np.array :sweat_smile: 
example:
 array(['5FG      ', 'the      ', 'either   ', 'Sincerely', 'cars     ',
       'From:    ', 'the      ', 'and\nthat ', 'old      '], dtype='<U9')]

fg-mindee replied:
I don't get how 1box label can have `\n` in it. Especially if we split it afterwards, 1 box = 2 text labels?
for your example, this would be more efficient:
```python
labels = [elt for word in txt.tolist() for elt in word.split()]
```

felixdittrich92 replied:
applied

fg-mindee commented on doctr/datasets/synthtext.py:
```diff
@@ -0,0 +1,85 @@
+
+class SynthText(VisionDataset):
+    """SynthText dataset from `"Synthetic Data for Text Localisation in Natural Images"
+    <https://arxiv.org/abs/1604.06646>`_.
```
I'd argue it's worth mentioning the repo https://github.com/ankush-me/SynthText or the website https://www.robots.ox.ac.uk/~vgg/data/scenetext/ as well

fg-mindee reviewed (changes requested):
Thanks, I added a few comments!

fg-mindee reviewed (approved):
Looks good, thanks!
